### DISCUSSION TITLE
Understand casync

### DISCUSSION THREAD

**Discussion by Sourabh1107** [rauc/rauc#794] [Q&A]
Hi,

I am a new to both RAUC and casync. I am trying to understand more about casync facility in RAUC and it would be great if we can have detailed documentation in the manual. Currently the little information is provided in the manual and its difficult for newbees to understand. Detailed section with examples would be helpful. I am interested to explore casync and implement it to do only application filesystem(AppFS) updates. 

P.S: I know its more of a question and discussion than an issue.

**Comment by jluebbe** [comment]
RAUC itself is mainly targeted at system updates (including changing the active A/B slot after writing the update). So app-only updates is not as easy as it would seem. I'm going to convert this to a discussion.

**Reply by Sourabh1107** [reply]
Hi @jluebbe thanks for the update. I understand your point. So basically RAUC is designed for system updates rather than application updates. Does casync is only used to reduce the size of the bundle? And I assume the other optimization is to save update time is to skip the installation of images to slots(slot skipping).

**Reply by jluebbe** [reply]
Many users of RAUC place the application in the rootfs, as this simplifies the update creation and installation. The trade-off is increased bundle size.

RAUC will can skip installing to slots if the image hash has not changed, independent of whether casync is used or not. (see https://rauc.readthedocs.io/en/latest/basic.html?highlight=skip#slot-status-and-skipping-slot-updates).

Using casync can help to reduce the download size, depending on how much of the image has changed from the previous version. As the bundle only contains the casync index (`.caibx` or `.caidx`), it will be very small. casync will then download only the required chunks during installation. Note that except in scenarios with very slow download speeds, using casync will likely *increase* installation time, as it needs to scan for reusable chunks before starting the download.

**Reply by Sourabh1107** [reply]
Hi @jluebbe thanks for the clarification. Basically I am trying to do only AppFS update. I completely understand that RAUC is designed for system updates but I am evaluating how we can modify this to accomplish only AppFS update. We already have RAUC setup for system updates(A/B update) and there could be other update systems which performs image-based delta updates but we are trying our best to stick with RAUC :-). Our whole idea is to copy the latest (current) rootfs of the active slot to local memory like SD card and whenever there is only appfs update, first check whether the rootfs of inactive slot(where update has to take place) has the latest rootfs(by matching hash sent along with update bundle. Update bundle consists of only hash of rootfs+appfs image). If hash doesn't match then first copy the current rootfs to inactive slot from the local memory and then perform the appfs update. And whenever there is a complete system update(both rootfs and appfs) copy the update to local memory. Metadata of current system will be sent to hawkbit before the update. Having the copy of rootfs in the local memory will also be useful in cases whenever if there any rootfs corruption in active slot.

I don't know whether we can actually implement it. Please correct me if I am missing something here and what could be the bottlenecks of this approach?

**Reply by Sourabh1107** [reply]
Hi @jluebbe @ejoerns what do you think of above approach? 

P.S: Sorry for hustling you guys. I am a Masters student working on embedded software updates and I want to implement this as part of my thesis work. I am open to any kind of suggestions.

**Reply by jluebbe** [reply]
The proper path forward depends on **why** you want to do an AppFS only update. Is it to reduce installation time, download size or something else?

Using casync can reuse the existing rootfs, reducing the download size, while avoiding a fixed dependency on which version is currently installed. It will just download the missing parts.

What you describe now seems to be similar to what @ejoerns suggested in https://github.com/rauc/rauc/issues/376#issuecomment-952787786: If you have a bundle which only contains a slot hash (instead of an image), check if the target is already the correct version (already implemented as slot skipping, see https://rauc.readthedocs.io/en/latest/reference.html#install-same), if not check if the correct hash is available in the active copy. If so, copy the contents.

What do you mean by 'local storage'? I'd say that all RAUC slots are normally on a local storage device (eMMC, SD, SSD, ...). So I'm not sure what you mean by 'copy the latest (current) rootfs of the active slot to local memory like SD card'. You already have the active root FS (in the root FS partition). Why not install the inactive side from there? Why do you need an additional copy?

Copying the rootfs is obviously safe if you use a read-only file system like squashfs, erofs or any dm-verity-protected image. Otherwise, if the root FS might be mounted read-write, you probably need to copy the contents as files (perhaps via rsync), as just copying the raw partition will lead to a corrupted copy.

**Reply by Sourabh1107** [reply]
Our goal is to roll out pure application updates on IoT devices without having to download a complete image (including RootFS). This reduces bundle size (less download overhead) and saves update time by not updating rootfs.

> Why not install the inactive side from there? Why do you need an additional copy?

I understand your point. Since RAUC slots are available on a local storage device there is no need to 'copy the latest (current) rootfs of the active slot to local memory like SD card'. Then, no need to do additional copy.

Scenario:

Active slot0:
RootFS A
Version: 1
AppFS A
Version: 1

Inactive slot1:
RootFS B
Version: 1
AppFS B
Version: 1

Trigger full system update via RAUC Bundle:
RAUC Bundle contents:
RootFS 
Version: 2
Hash:XYZ
AppFS
Version:2
Hash: ABC

After update and reboot:
Inactive slot0:
RootFS A
Version: 1
AppFS A
Version: 1

Active slot1:
RootFS B
Version: 2
AppFS B
Version: 2

Next time, I want to roll out only application filesystem update to reduce bundle and download size
RAUC Bundle contents:
Hash: XYZ (Hash of RootFS V2)
AppFS 
Version: 3
Hash: XXX

Since now the inactive slot is slot0 and it contains RootFS A V1. But the active slot is slot1 with RootFS B V2. So, before the update, RootFS B V2(from active slot) should be copied to slot0 so that the compatibility of the AppFS image to be installed with the one in the RootFS in the new slot is ensured. So, after the update the active slot should look like:
Active slot0:
RootFS A
Version: 2
AppFS A
Version: 3

In order to trigger an AppFS update in Hawkbit, metadata will be sent to hawkbit E.g. the number and hashes of the locally cached bundles.
Based on this metadata, it should then be possible to roll out through rules, E.g. if RootFS with hash XYZ is available locally, do AppFS only update.

Does slot skipping reduces the bundle and download size? As far I understand, it only optimizes the update time but update bundle still consists of rootfs image and appfs image.

P.S: I have tried my best to describe our use case. Please let me know if there any irregularities. 

![image](https://user-images.githubusercontent.com/93142990/141803374-d0edc6a8-92a9-490e-b590-ffb8df323b57.png)
Example scenario

**Reply by Sourabh1107** [reply]
Hi @ejoerns @jluebbe  what do you think of above scenario? Is this valid scenario which can be tackled with appfs only updates?
